File: IDATT2106 Systemutvikling 2 med smidig prosjekt/backend/src/main/java/com/ntnu/gidd/security/token/RawJwtAccessToken.java
package com.ntnu.gidd.security.token;

import com.ntnu.gidd.exception.JwtExpiredTokenException;
import io.jsonwebtoken.*;
import lombok.AllArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.authentication.BadCredentialsException;


/**
 * Class for unparsed raw jwt access tokens.
 */
@Slf4j
@AllArgsConstructor
public class RawJwtAccessToken implements JwtToken {

    private String token;

    /**
     * Parse and validate jwt signature
     * @return the jwt claims.
     */
    public Jws<Claims> parseClaims(String signingKey) {
        try {
            return Jwts.parser()
                    .setSigningKey(signingKey)
                    .parseClaimsJws(this.token);
        } catch (UnsupportedJwtException | MalformedJwtException | IllegalArgumentException | SignatureException ex) {
            handleInvalidJwtToken(ex);
        } catch (ExpiredJwtException ex) {
            handleExpiredJwtToken(ex);
        }

        return null;
    }

    private void handleInvalidJwtToken(RuntimeException ex) {
        log.error("Invalid jwt token: {}", token, ex);
        throw new BadCredentialsException("Invalid jwt token: ", ex);
    }

    private void handleExpiredJwtToken(ExpiredJwtException ex) {
        log.info("Jwt token is expired", ex);
        throw new JwtExpiredTokenException(ex.getMessage(), ex);
    }

    @Override
    public String getToken() {
        return token;
    }
}

File: IDATT2105 Full-stack applikasjonsutvikling/Rombestilling/backend/src/main/kotlin/ntnu/idatt2105/security/token/RawJwtAccessToken.kt
package ntnu.idatt2105.security.token

import io.jsonwebtoken.*
import ntnu.idatt2105.security.exception.JwtExpiredTokenException
import org.springframework.security.authentication.BadCredentialsException
import java.security.SignatureException

data class RawJwtAccessToken(private var token: String) : JwtToken {

    fun parseClaims(signingKey: String?): Jws<Claims>? {
        try {
            return Jwts.parserBuilder()
                    .setSigningKey(signingKey)
                    .build()
                    .parseClaimsJws(token)
        } catch (ex: UnsupportedJwtException) {
            handleInvalidJwtToken(ex)
        } catch (ex: MalformedJwtException) {
            handleInvalidJwtToken(ex)
        } catch (ex: IllegalArgumentException) {
            handleInvalidJwtToken(ex)
        } catch (ex: SignatureException) {
            handleInvalidJwtToken(ex)
        } catch (ex: ExpiredJwtException) {
            handleExpiredJwtToken(ex)
        }
        return null
    }

    private fun handleInvalidJwtToken(ex: RuntimeException) {
        throw BadCredentialsException("Invalid jwt token: ", ex)
    }
    private fun handleInvalidJwtToken(ex: SignatureException) {
        throw BadCredentialsException("Invalid jwt token: ", ex)
    }

    private fun handleExpiredJwtToken(ex: ExpiredJwtException) {
        throw JwtExpiredTokenException(ex.message, ex)
    }

    override fun getToken(): String {
        return token
    }
}
